metrics/cloudwatch2/cloudwatch2.go

```diff
@@ -0,0 +1,249 @@
+// Callers must ensure that regular calls to Send are performed, either
+// manually or with one of the helper methods.
+func New(namespace string, svc cloudwatchiface.CloudWatchAPI, options ...option) *CloudWatch {
+	cw := &CloudWatch{
+		sem:                   nil, // set below
+		namespace:             namespace,
+		svc:                   svc,
+		counters:              lv.NewSpace(),
+		numConcurrentRequests: 10,
+		logger:                log.NewLogfmtLogger(os.Stderr),
+	}
+
+	for _, optFunc := range options {
+		optFunc(cw)
+	}
+
+	cw.sem = make(chan struct{}, cw.numConcurrentRequests)
+
+	return cw
+}
+
+// NewCounter returns a counter. Observations are aggregated and emitted once
+// per write invocation.
+func (cw *CloudWatch) NewCounter(name string) metrics.Counter {
+	return &Counter{
+		name: name,
+		obs:  cw.counters.Observe,
+	}
+}
+
+// NewGauge returns an gauge. Under the covers, there is no distinctions
+// in CloudWatch for how Counters/Histograms/Gauges are reported, so this
+// just wraps a cloudwatch2.Counter.
+func (cw *CloudWatch) NewGauge(name string) metrics.Gauge {
+	return convert.NewCounterAsGauge(cw.NewCounter(name))
+}
+
+// NewHistogram returns a histogram. Under the covers, there is no distinctions
+// in CloudWatch for how Counters/Histograms/Gauges are reported, so this
+// just wraps a cloudwatch2.Counter.
+func (cw *CloudWatch) NewHistogram(name string) metrics.Histogram {
+	return convert.NewCounterAsHistogram(cw.NewCounter(name))
+}
+
+// WriteLoop is a helper method that invokes Send every time the passed
+// channel fires. This method blocks until the channel is closed, so clients
+// probably want to run it in its own goroutine. For typical usage, create a
+// time.Ticker and pass its C channel to this method.
+func (cw *CloudWatch) WriteLoop(c <-chan time.Time) {
+	for range c {
+		if err := cw.Send(); err != nil {
+			cw.logger.Log("during", "Send", "err", err)
+		}
+	}
+}
+
+// Send will fire an API request to CloudWatch with the latest stats for
+// all metrics. It is preferred that the WriteLoop method is used.
+func (cw *CloudWatch) Send() error {
+	cw.mtx.RLock()
+	defer cw.mtx.RUnlock()
+	now := time.Now()
+
+	var datums []cloudwatch.MetricDatum
+
+	cw.counters.Reset().Walk(func(name string, lvs lv.LabelValues, values []float64) bool {
+		datums = append(datums, cloudwatch.MetricDatum{
+			MetricName:      aws.String(name),
+			Dimensions:      makeDimensions(lvs...),
+			StatisticValues: stats(values),
+			Timestamp:       aws.Time(now),
+		})
+		return true
+	})
+
+	var batches [][]cloudwatch.MetricDatum
+	for len(datums) > 0 {
+		var batch []cloudwatch.MetricDatum
+		lim := len(datums)
+		if lim > maxConcurrentRequests {
+			lim = maxConcurrentRequests
+		}
+		batch, datums = datums[:lim], datums[lim:]
+		batches = append(batches, batch)
+	}
+
+	var errors = make(chan error, len(batches))
+	for _, batch := range batches {
+		go func(batch []cloudwatch.MetricDatum) {
```

Comment:
the goroutines could run as an [errorgroup](https://godoc.org/golang.org/x/sync/errgroup)

Reply:
Oooh, cool! I hadn't yet added `errorgroup` to my toolbelt. Added!